vavr/src/main/java/org/jdbi/v3/vavr/VavrOptionMapper.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.vavr;
+
+import io.vavr.control.Option;
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.mapper.ColumnMapper;
+import org.jdbi.v3.core.mapper.ColumnMapperFactory;
+import org.jdbi.v3.core.mapper.NoSuchMapperException;
+import org.jdbi.v3.core.statement.StatementContext;
+
+import java.lang.reflect.Type;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+
+import static org.jdbi.v3.core.generic.GenericTypes.getErasedType;
+
+public class VavrOptionMapper<T> implements ColumnMapper<Option<T>> {
+
+    private final Type type;
+
+    private VavrOptionMapper(Type type) {
+        this.type = type;
+    }
+
+    public static ColumnMapper<?> of(Type type) {
+        return new VavrOptionMapper(type);
+    }
+
+    public static ColumnMapperFactory factory() {
+        return (type, config) -> {
+            Class<?> rawType = getErasedType(type);
+            if (rawType == Option.class) {
+                return Optional.of(VavrOptionMapper.of(type));
```

Comment:
I think this lambda should be where we unwrap the `Option<T>` generic type instead of `map()`, so that `of(type)` receives the element type `T` instead of `Option<T>`.

Maybe rename the `of()` method to `ofElement()` to remove ambiguity?

One side benefit of doing the `findGenericParameter(...).orElseThrow(...)` here instead of in `map()` is that this code executes _before_ the database statement is executed.

Reply:
Can do, but is wanted to stay consistent with the `OptionalMapper` implementation. Are you aware that this code is just analogue to the core code?

Reply:
Yikes. I did not know that. Thankfully, that class is nonpublic. Sheesh..

Reply:
Would you mind hoisting that generic parameter check in `OptionalMapper.map()` into `BuiltInMapperFactory.build()` while you're at it?

Reply:
done. furthermore i removed the `of` method anyway. i dont see a usecase where someone would need a public access to create a new `OptionMapper` - same goes for package private visibility of the factory method.

If you agree, same could be said for the `OptionalMapper` class

I renamed the variable to make it clear that the nested type is under consideration

Reply:
With the `factory()` method hidden package private and the constructor private, does it make sense for `VavrOptionMapper` to be public at all? I think it's okay for now to just expose the feature through the plugin.

Reply:
Thinking about it - I will leave the factory method public and the constructor private. It can make sense to allow public access to the factory if you only want jdbi customized with selected features of the plugin and not use the `VavrPlugin` class. I will check if this holds for all features that are installed via VavrPlugin class

The constructor however will remain private since it makes no sense to me to register mappers of Option<T> with only handpicked T.

You agree with this?

Reply:
I think we're in agreement on visibility:

* Reduce the visibility of `OptionMapper` to package private
* Reduce the visibility of `OptionalMapper.of()` to package private. Since the class is already package private, reducing the method visibility doesn't hurt anything.
* (Optional, if you care to take the time) Refactor `OptionalMapper` so it accepts the nested type instead of `Optional<Something>`--similar to what you did with `VavrOptionMapper`.

Reply:
Are we in agreement? :) 

I just proposed to leave the visibility of `VavrOptionMapper` public to enable client calls to the factory method. that way the user can register parts of the plugin with JDBI directly instead of using the `VavrPlugin` class.

Then again I am not sure if a JDBI plugin is expected to be partly installable or only as a whole - I don't know if there is any guideline here

Reply:
There aren't any plugin guidelines yet--we're still learning how people will use this. We'll document best practices as we discover them.

As for the public/private thing, I can see both sides of it. If we kept it public, we'd want to make the factory method also public--right now it's package private.

However I'd rather not introduce permanent public API based on speculation. We don't know yet whether people will want access to the mapper, or will be content with just the plugin.

If we make it private now, we can always make it public later if users ask for it. But if we make it public now and realize later that it was a mistake, we can't make it private without breaking compatibility.

So my preference is to make the mapper private, and see how far we get with just the plugin.